utils/fill_zero.go

```diff
@@ -0,0 +1,111 @@
+func NewFillZeroCmd(trace bool) *FillZero {
+	return &FillZero{}
+}
+
+func (z *FillZero) Fill(ctx context.Context, logicalName string) error {
+	log.Printf("Start overwriting with zeros %s", logicalName)
+	partitionPath := logicalName // example /dev/sdb
+
+	// Buffer size (in bytes)
+	bufferSize := 4096 // to use 1MB = 1024 * 1024
+
+	// Write open
+	file, err := os.OpenFile(partitionPath, os.O_WRONLY, FileMode644)
+	if err != nil {
+		log.Println("Failed to open :"+logicalName, err)
+		return err
+	}
+	defer file.Close()
+
+	// Get disk or partition size
+	whence := 2 // 2 means relative to the end, check Seek function
+	partitionSize, err := file.Seek(0, whence)
+	if err != nil {
+		log.Println(err)
+		return err
+	}
+
+	log.Printf("%s | Size in bytes: %d \n", partitionPath, partitionSize)
+
+	// Create a buffer fill with zeroes
+	buffer := make([]byte, bufferSize)
+
+	var totalBytesWritten int64
+	var bytesSinceLastPrint int64
+	start := time.Now()
+
+	// rewind cassette tape ;=)
+	whence = 0 // 0 means relative to the origin of the file, check Seek function
+	_, err = file.Seek(0, whence)
+	if err != nil {
+		log.Println(err)
+		return err
+	}
+
+	// Writing zeroes loop
+	bytesSinceLastPrint = 0
+	for totalBytesWritten < partitionSize {
+		bytesWritten, err := file.Write(buffer)
+		if err != nil {
+			if strings.Contains(err.Error(), "no space left on device") { // syscall.ENOSPC
+				// Check if the last write loop didn't write all the buffer size, the reason is partitionSize % bufferSize is not 0
+				if totalBytesWritten+int64(bytesWritten) == partitionSize {
+					log.Println("we have reached the end of the devices.")
+					return nil
+				} else {
+					log.Println("we have a problem with the write to disk: ", err)
+				}
+			} else {
+				// Other errors
+				log.Println("failed to write to disk: ", err)
+			}
+			return err
+		}
+		totalBytesWritten += int64(bytesWritten)
+		bytesSinceLastPrint += int64(bytesWritten)
+
+		// Print progress each 10 seconds
+		if time.Since(start) >= 10*time.Second {
+			// Calculate progress and ETA
+			progress := float64(totalBytesWritten) / float64(partitionSize) * TotalPercentage
+			elapsed := time.Since(start).Seconds()
+			speed := float64(bytesSinceLastPrint) / elapsed                                   // Speed in bytes per second
+			remainingSeconds := (float64(partitionSize) - float64(totalBytesWritten)) / speed // Remaining time in seconds
+			remainingHours := float64(remainingSeconds / SecondsPerHour)
+			mbPerSecond := speed / OneMBinBytes
+			log.Printf("%s | Progress: %.2f%% | Speed: %.2f MB/s | Estimated time left: %.2f hour(s)\n", partitionPath, progress, mbPerSecond, remainingHours)
+			start = time.Now()
+			bytesSinceLastPrint = 0
+		}
+	}
+
+	log.Printf("%s has been completely overwritten with zeros", partitionPath)
+
+	return nil
+}
+
+func (z *FillZero) WipeDisk(ctx context.Context, logicalName string) error {
+	return z.Fill(ctx, logicalName)
+}
```

Comment:
I wanted to test this out and see what happens if we get ENOSPC and if we could use `errors.Is(err, syscall.ENOSPC)` (yes that would have worked) and kind of got distracted/carried away cleaning some things up. But, I think this comes out cleaner now. Gomnd is annoying most times and this time is no different. I also dropped the ENOSPC by just writing `min(len(buffer), bytesRemaining)` instead, having `bytesRemaining` turned out to be pretty useful for the progress report too.

```suggestion
//noline:gomnd // the magic numbers here are fine
func (z *FillZero) WipeDisk(ctx context.Context, path string) error {
	fmt.Println("Starting zero-fill of", path)

	// Write open
	file, err := os.OpenFile(path, os.O_WRONLY, 0)
	if err != nil {
		log.Println(fmt.Errorf("%w", err))
		return err
	}
	defer file.Close()

	// Get disk or partition size
	partitionSize, err := file.Seek(0, io.SeekEnd)
	if err != nil {
		log.Println(err)
		return err
	}
	fmt.Printf("%s | Size: %dB\n", path, partitionSize)

	_, err = file.Seek(0, io.SeekStart)
	if err != nil {
		return err
	}

	var bytesSinceLastPrint int
	var totalBytesWritten int
	buffer := make([]byte, 4096)
	start := time.Now()

	for bytesRemaining := int(partitionSize); bytesRemaining > 0; {
		l := min(len(buffer), bytesRemaining)
		bytesWritten, err := file.Write(buffer[:l])
		if err != nil {
			return err
		}

		totalBytesWritten += bytesWritten
		bytesSinceLastPrint += bytesWritten
		bytesRemaining -= bytesWritten

		// Print progress report every 10 seconds and when done
		if bytesRemaining == 0 || time.Since(start) >= 10*time.Second {
			progress := float64(totalBytesWritten) / float64(partitionSize) * 100
			elapsed := time.Since(start)
			rate := int(int64(bytesSinceLastPrint) / elapsed.Nanoseconds())
			remaining := time.Duration(bytesRemaining / rate)

			mbPerSecond := float64(time.Duration(rate)*time.Second) / (1 * 1024 * 1024)
			fmt.Printf("%s | Progress: %6.2f%% | Speed: %.2f MB/s | Estimated time left: %5.2fh\n",
				path, progress, mbPerSecond, remaining.Hours())

			start = time.Now()
			bytesSinceLastPrint = 0
		}
	}

	file.Sync()
	return nil
}
```

Reply:
Here's an example run where I printed every 10ms and changed .Hours() -> .Seconds()
```
Starting zero-fill of /dev/loop0
/dev/loop0 | Size: 1073741824B
/dev/loop0 | Progress:   1.71% | Speed: 953.67 MB/s | Estimated time left:  1.06h
/dev/loop0 | Progress:   3.45% | Speed: 953.67 MB/s | Estimated time left:  1.04h
/dev/loop0 | Progress:   5.18% | Speed: 953.67 MB/s | Estimated time left:  1.02h
/dev/loop0 | Progress:   6.94% | Speed: 953.67 MB/s | Estimated time left:  1.00h
/dev/loop0 | Progress:   8.68% | Speed: 953.67 MB/s | Estimated time left:  0.98h
/dev/loop0 | Progress:  10.45% | Speed: 953.67 MB/s | Estimated time left:  0.96h
/dev/loop0 | Progress:  12.21% | Speed: 953.67 MB/s | Estimated time left:  0.94h
/dev/loop0 | Progress:  13.98% | Speed: 953.67 MB/s | Estimated time left:  0.92h
/dev/loop0 | Progress:  15.74% | Speed: 953.67 MB/s | Estimated time left:  0.90h
/dev/loop0 | Progress:  17.50% | Speed: 953.67 MB/s | Estimated time left:  0.89h
/dev/loop0 | Progress:  19.27% | Speed: 953.67 MB/s | Estimated time left:  0.87h
/dev/loop0 | Progress:  21.03% | Speed: 953.67 MB/s | Estimated time left:  0.85h
/dev/loop0 | Progress:  22.81% | Speed: 953.67 MB/s | Estimated time left:  0.83h
/dev/loop0 | Progress:  24.57% | Speed: 953.67 MB/s | Estimated time left:  0.81h
/dev/loop0 | Progress:  26.34% | Speed: 953.67 MB/s | Estimated time left:  0.79h
/dev/loop0 | Progress:  28.11% | Speed: 953.67 MB/s | Estimated time left:  0.77h
/dev/loop0 | Progress:  29.86% | Speed: 953.67 MB/s | Estimated time left:  0.75h
/dev/loop0 | Progress:  31.62% | Speed: 953.67 MB/s | Estimated time left:  0.73h
/dev/loop0 | Progress:  33.39% | Speed: 953.67 MB/s | Estimated time left:  0.72h
/dev/loop0 | Progress:  35.15% | Speed: 953.67 MB/s | Estimated time left:  0.70h
/dev/loop0 | Progress:  36.88% | Speed: 953.67 MB/s | Estimated time left:  0.68h
/dev/loop0 | Progress:  38.62% | Speed: 953.67 MB/s | Estimated time left:  0.66h
/dev/loop0 | Progress:  40.37% | Speed: 953.67 MB/s | Estimated time left:  0.64h
/dev/loop0 | Progress:  42.11% | Speed: 953.67 MB/s | Estimated time left:  0.62h
/dev/loop0 | Progress:  43.88% | Speed: 953.67 MB/s | Estimated time left:  0.60h
/dev/loop0 | Progress:  45.64% | Speed: 953.67 MB/s | Estimated time left:  0.58h
/dev/loop0 | Progress:  47.40% | Speed: 953.67 MB/s | Estimated time left:  0.56h
/dev/loop0 | Progress:  49.15% | Speed: 953.67 MB/s | Estimated time left:  0.55h
/dev/loop0 | Progress:  50.92% | Speed: 953.67 MB/s | Estimated time left:  0.53h
/dev/loop0 | Progress:  52.69% | Speed: 953.67 MB/s | Estimated time left:  0.51h
/dev/loop0 | Progress:  54.46% | Speed: 953.67 MB/s | Estimated time left:  0.49h
/dev/loop0 | Progress:  56.23% | Speed: 953.67 MB/s | Estimated time left:  0.47h
/dev/loop0 | Progress:  58.01% | Speed: 953.67 MB/s | Estimated time left:  0.45h
/dev/loop0 | Progress:  59.79% | Speed: 953.67 MB/s | Estimated time left:  0.43h
/dev/loop0 | Progress:  61.56% | Speed: 953.67 MB/s | Estimated time left:  0.41h
/dev/loop0 | Progress:  63.33% | Speed: 953.67 MB/s | Estimated time left:  0.39h
/dev/loop0 | Progress:  65.10% | Speed: 953.67 MB/s | Estimated time left:  0.37h
/dev/loop0 | Progress:  66.86% | Speed: 953.67 MB/s | Estimated time left:  0.36h
/dev/loop0 | Progress:  68.60% | Speed: 953.67 MB/s | Estimated time left:  0.34h
/dev/loop0 | Progress:  70.43% | Speed: 953.67 MB/s | Estimated time left:  0.32h
/dev/loop0 | Progress:  72.24% | Speed: 953.67 MB/s | Estimated time left:  0.30h
/dev/loop0 | Progress:  74.05% | Speed: 953.67 MB/s | Estimated time left:  0.28h
/dev/loop0 | Progress:  75.88% | Speed: 953.67 MB/s | Estimated time left:  0.26h
/dev/loop0 | Progress:  77.70% | Speed: 953.67 MB/s | Estimated time left:  0.24h
/dev/loop0 | Progress:  79.54% | Speed: 953.67 MB/s | Estimated time left:  0.22h
/dev/loop0 | Progress:  81.37% | Speed: 953.67 MB/s | Estimated time left:  0.20h
/dev/loop0 | Progress:  83.20% | Speed: 953.67 MB/s | Estimated time left:  0.18h
/dev/loop0 | Progress:  85.04% | Speed: 953.67 MB/s | Estimated time left:  0.16h
/dev/loop0 | Progress:  86.86% | Speed: 953.67 MB/s | Estimated time left:  0.14h
/dev/loop0 | Progress:  88.69% | Speed: 953.67 MB/s | Estimated time left:  0.12h
/dev/loop0 | Progress:  90.53% | Speed: 953.67 MB/s | Estimated time left:  0.10h
/dev/loop0 | Progress:  92.35% | Speed: 953.67 MB/s | Estimated time left:  0.08h
/dev/loop0 | Progress:  94.19% | Speed: 953.67 MB/s | Estimated time left:  0.06h
/dev/loop0 | Progress:  96.01% | Speed: 953.67 MB/s | Estimated time left:  0.04h
/dev/loop0 | Progress:  97.84% | Speed: 953.67 MB/s | Estimated time left:  0.02h
/dev/loop0 | Progress:  99.65% | Speed: 953.67 MB/s | Estimated time left:  0.00h
/dev/loop0 | Progress: 100.00% | Speed: 953.67 MB/s | Estimated time left:  0.00h
```

Reply:
Implemented, used old print progress due to division by zero